pennylane/resource/specs.py

```diff
@@ -12,19 +12,165 @@
+                # In the case of a broad exception, we don't want the `qml.specs` transform
+                # to fail. Instead, we simply indicate that the number of gradient executions
+                # is not supported for the reason specified.
+                info["num_gradient_executions"] = f"NotSupported: {str(e)}"
+        else:
+            info["gradient_fn"] = gradient_fn
+
+        infos.append(info)
+
+    return infos[0] if len(infos) == 1 else infos
+
+
+# NOTE: Some information is missing from specs_qjit compared to specs_qnode
+def _specs_qjit(qjit, level, compute_depth, *args, **kwargs) -> SpecsDict:  # pragma: no cover
+    # pylint: disable=import-outside-toplevel
+    # Have to import locally to prevent circular imports as well as accounting for Catalyst not being installed
+    # Integration tests for this function are within the Catalyst frontend tests, it is not covered by unit tests
+
+    from catalyst.jit import QJIT
+
+    from ..devices import NullQubit
+
+    # TODO: Determine if its possible to have batched QJIT code / how to handle it
+
+    if not isinstance(qjit.original_function, qml.QNode):
+        raise ValueError("qml.specs can only be applied to a QNode or qjit'd QNode")
+
+    original_device = qjit.device
+    info = SpecsDict()
+
+    if level != "device":
+        raise NotImplementedError(f"Unsupported level argument '{level}' for QJIT'd code.")
+
+    # When running at the device level, execute on null.qubit directly with resource tracking,
+    # which will give resource usage information for after all compiler passes have completed
+
+    # TODO: Find a way to inherit all devices args from input
+    spoofed_dev = NullQubit(
+        target_device=original_device,
+        wires=original_device.wires,
+        shots=original_device.shots,
+        track_resources=True,
+        resources_filename=_RESOURCE_TRACKING_FILEPATH,
+        compute_depth=compute_depth,
+    )
+
+    new_qnode = qjit.original_function.update(device=spoofed_dev)
+    new_qjit = QJIT(new_qnode, copy.copy(qjit.compile_options))
+
+    if os.path.exists(_RESOURCE_TRACKING_FILEPATH):
+        # TODO: Warn that something has gone wrong here
+        os.remove(_RESOURCE_TRACKING_FILEPATH)
+
+    # Execute on null.qubit with resource tracking
+    new_qjit(*args, **kwargs)
+
+    with open(_RESOURCE_TRACKING_FILEPATH, "r", encoding="utf-8") as f:
+        resource_data = json.load(f)
+
+    info["resources"] = Resources(
+        num_wires=resource_data["num_wires"],
+        num_gates=resource_data["num_gates"],
+        gate_types=defaultdict(int, resource_data["gate_types"]),
+        gate_sizes=defaultdict(int, {int(k): v for (k, v) in resource_data["gate_sizes"].items()}),
+        depth=resource_data["depth"],
+        shots=qjit.original_function.shots,  # TODO: Can this ever be overriden during compilation?
+    )
+
+    # print(resource_data)
+    os.remove(_RESOURCE_TRACKING_FILEPATH)
```

Comment:
Should we use a `try-finally` block to make sure we are always cleaning up the file?